src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1498,81 +1441,77 @@ std::vector<MergeTreeData::LoadPartResult> MergeTreeData::loadDataPartsFromDisk(
+                parts_futures.push_back(runner(
+                    [&, part = std::move(current_part)]()
+                    {
+                        /// Pass a separate mutex to guard the set of parts, because this lambda
+                        /// is called concurrently but with already locked @data_parts_mutex.
+                        auto res = loadDataPartWithRetries(
+                            part->info, part->name, part->disk,
+                            DataPartState::Active, part_loading_mutex, loading_parts_initial_backoff_ms,
+                            loading_parts_max_backoff_ms, loading_parts_max_tries);
+
+                        part->is_loaded = true;
+                        bool is_active_part = res.part->getState() == DataPartState::Active;
+
+                        /// If part is broken or duplicate or should be removed according to transaction
+                        /// and it has any covered parts then try to load them to replace this part.
+                        if (!is_active_part && !part->children.empty())
                         {
-                            // Try random steal tasks from the next thread
-                            std::uniform_int_distribution<size_t> distribution(0, remaining_thread_parts.size() - 1);
-                            auto it = remaining_thread_parts.begin();
-                            std::advance(it, distribution(thread_local_rng));
-                            thread_idx = *it;
+                            std::lock_guard lock{part_select_mutex};
+                            for (const auto & [_, node] : part->children)
+                                parts_to_load.push_back(node);
                         }
 
-                        auto & thread_parts = threads_parts[thread_idx];
-                        thread_part = thread_parts.back();
-                        thread_parts.pop_back();
-                        if (thread_parts.empty())
-                            remaining_thread_parts.erase(thread_idx);
-                    }
+                        {
+                            std::lock_guard lock(part_loading_mutex);
+                            loaded_parts.push_back(std::move(res));
+                        }
+                    }, 0
+                ));
+            }
 
-                    /// Pass a separate mutex to guard the set of parts, because this lambda
-                    /// is called concurrently but with already locked @data_parts_mutex.
-                    auto res = loadDataPartWithRetries(
-                        thread_part->info, thread_part->name, thread_part->disk,
-                        DataPartState::Active, part_loading_mutex, loading_parts_initial_backoff_ms,
-                        loading_parts_max_backoff_ms, loading_parts_max_tries);
+            /// Wait for all scheduled tasks
+            for (auto & future: parts_futures)
+                future.wait();
```

Comment:
we never call `get()` on futures, so exceptions will be ignored. is it okay here?

Reply:
Maybe call `get()` here and keep `wait()` in catch block?
```suggestion
                future.get();
```

Reply:
Fixed!